app/objects/secondclass/c_link.py

```diff
@@ -218,31 +218,57 @@ async def _load_module(module_type, module_info):
 
     async def _create_relationships(self, relationships, operation):
         for relationship in relationships:
-            await self._save_fact(operation, relationship.source, relationship.score)
-            await self._save_fact(operation, relationship.target, relationship.score)
+            relationship.origin = operation.id if operation else self.id
+            await self._save_fact(operation, relationship.source, relationship.score, relationship.shorthand)
+            await self._save_fact(operation, relationship.target, relationship.score, relationship.shorthand)
             if all((relationship.source.trait, relationship.edge)):
+                knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+                await knowledge_svc_handle.add_relationship(relationship)
                 self.relationships.append(relationship)
 
-    async def _save_fact(self, operation, fact, score):
-        all_facts = operation.all_facts() if operation else self.facts
-        if all([fact.trait, fact.value]) and await self._is_new_fact(fact, all_facts):
-            self.facts.append(Fact(trait=fact.trait, value=fact.value, score=score, collected_by=self.paw,
-                                   technique_id=self.ability.technique_id))
+    async def _save_fact(self, operation, fact, score, relationship):
+        knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+        all_facts = await operation.all_facts() if operation else self.facts
+        source = operation.id if operation else self.id
+        if all([fact.trait, fact.value]):
+            if await self._is_new_fact(fact, all_facts):
```

Comment:
Most of this is already offered by the Knowledge Service in some capacity, and the rest wouldn't be hard to add. While I like the idea of moving most of this stuff over there, I am a little hesitant since doing so will mean a complete breakage of the old style of fact storage. This method maintains backwards compatibility for anyone/anything that references a link's facts directly (which I think a few custom planners do). At some point we can break that and move this all into the Knowledge Service, which would be nice and clean, but unless everyone wants to do that now, I think it might be better to hold off a little while.

Reply:
Gotcha.